doc/read-the-docs-site/adr/0003-sharing-prod-debugging-cek.md

```diff
@@ -0,0 +1,117 @@
+```
+
+We enter either modes via `enterComputeCek`, which takes an extra `Bool` than our current implementation, to indicate whether we are in debugging mode or not:
+
+```haskell
+enterComputeCek 
+    :: forall uni fun s
+    . (Ix fun, PrettyUni uni fun, GivenCekReqs uni fun s)
+    => Bool
+    -> Context uni fun
+    -> Closure uni fun
+    -> CekM uni fun s (CekState uni fun)
+enterComputeCek debug ctx (Closure term env) =
+    computeCekStep (toWordArray 0) NoFrame (Closure term Env.empty) where
+    
+    computeCek
+        :: WordArray -- for costing
+        -> Context uni fun
+        -> Closure uni fun
+        -> CekM uni fun s (CekState uni fun)
+    computeCek = if debug then computeCekDebug else computeCekStep
+    {-# NOINLINE computeCek #-}  -- Making sure the `if` is only evaluated once.
+
+    -- in debugging mode, immediately returns the current `CekState` and halts execution. Debugging mode details to be worked out.
+    computeCekDebug 
+        :: WordArray
+        -> Context uni fun
+        -> Closure uni fun
+        -> CekM uni fun s (CekState uni fun) 
+    computeCekDebug budget ctx (Closure term env) = 
+        pure $ Computing budget ctx (Closure term env)
+
+    -- in production mode, recursively calls `computeCek`, and thus `computeCekStep`, similar to our current production machine.
+    computeCekStep 
+        :: WordArray
+        -> Context uni fun
+        -> Closure uni fun
+        -> CekM uni fun s (CekState uni fun) -- the return type is `CekState` instead of a term.
+    computeCekStep unbudgetedSteps ctx (Closure (Force _ body) env) = do -- exactly like in current prod
+        !unbudgetedSteps' <- stepAndMaybeSpend BForce unbudgetedSteps -- update costs
+        computeCek unbudgetedSteps' (FrameForce ctx) (Closure body env) -- compute again with updated costs and ctx
+    <other_clauses>
+    
+    -- details of `forceEvaluate`, `applyEvaluate` etc to be worked out.
+
+    -- similarly for the returning step
+
+    returnCek = if debug then returnCekDebug else returnCekStep
+    {-# NOINLINE returnCek #-}
+    
+    returnCekDebug = ...
+
+    
+    returnCekStep 
+        :: forall uni fun s
+        . (PrettyUni uni fun, GivenCekReqs uni fun s)
+        => WordArray
+        -> Context uni fun
+        -> CekValue uni fun
+        -> CekM uni fun s (CekState uni fun) -- return a state instead of a term
+    returnCekStep !unbudgetedSteps NoFrame val = do
+    spendAccumulatedBudget unbudgetedSteps
+    pure $ Terminating $ dischargeCekValue val --wrap the term in the `Terminating` constructor when returning the term.
+    <other_clauses>
+```
+
+Because when we are not debugging, we are still using basically the same code as our current implementation, the performance should not be affected by much. (Given that the machine is tail-recursive, the additional wrapping of the returned term in the `Terminating` constructor will affect performance in a negligible way.)
```

Comment:
This doesn't seem like basically the same code to me: previously `computeCek` performs the whole evaluation in a tail recursion, and now the computation is broken into individual steps, and it may need to allocate `CekState` and immediately garbage-collect it at every step (although perhaps GHC can optimize that away? I don't know). Have you run the benchmark on the proposed approach?

Reply:
I think the example isn't quite clear. The idea is that we will (through type-level trickery) end up with two versions of the machine, _specialised_ to the two values of `debug`. The one that is specialized to `False` should produce similar code, since it just tail-calls itself, and only actually constructs a `CekState` at the very end.

Reply:
@zliu41 This hasn't been prototyped yet so no I haven't run the benchmark on the proposed approach. I could do so though if you think I should. Let me know.

@michaelpj it's true, I didn't include the whole story. I'll add that as best I can.